apiserver/facades/agent/uniter/uniter.go

```diff
@@ -3318,23 +3335,29 @@ func (u *UniterAPI) CommitHookChanges(args params.CommitHookChangesArgs) (params
 			continue
 		}
 
-		if !canAccess(unitTag) {
+		if !canAccessUnit(unitTag) {
 			res[i].Error = common.ServerError(common.ErrPerm)
 			continue
 		}
 
-		res[i].Error = common.ServerError(u.commitHookChangesForOneUnit(unitTag, arg, canAccess))
+		res[i].Error = common.ServerError(u.commitHookChangesForOneUnit(unitTag, arg, canAccessUnit, canAccessApp))
```

Comment:
The `canAccessApp` is used by `SetPodSpec` (but required only when changing the spec) and is a bit tricky because we are authenticated as a unit and not the app.

The `canAccessUnit` check is performed once at the outer loop but it is also used by the code that updates the relation data [here](https://github.com/juju/juju/blob/a5d64faf556928337d4c73b12ed581d2995d31ff/apiserver/facades/agent/uniter/uniter.go#L1772). As the goal of the PR was to run things transactionally and not refactor the entire facade, I opted to keep the existing code and just thread the access-checkers through the calls.